File: 2018summer/07.12/solutions_2015/carpets/submissions/accepted/Thorsten.java
import java.util.*;
import java.math.*;

public class Thorsten {
    public static int W;
    public static int H;
    public static int c;
    public static int[] a = new int[9];
    public static int[] w = new int[9];
    public static int[] h = new int[9];

    public static boolean[][] buf = new boolean[101][]; // whether occupied by carpet

    public static boolean tryFill(int x, int y) {
        while (true) {
            // do this...
            if (x >= W) { // go to next row
                x = 0;
                y++;
            }
            if (y >= H) {
                return true;
            }
            // until we find a free cell
            if (buf[x][y] == false) break;
            x++;
        }
        // try all the carpets
        for (int i = 0; i < c; i++) {
            if (a[i] <= 0) continue;
            // does it fit at all?
            int wi = w[i], hi = h[i];
            for (int xyz = 0; xyz < 2; xyz++) {
                int tmp = wi;
                wi = hi;
                hi = tmp;
                if (x + wi > W) continue;
                if (y + hi > H) continue;
                boolean allfree = true;
                // we only need to check the [dx + x][y] points. because if so,
                // any [dx + x][y + dy] will be free as well.
                for (int dx = 0; dx < wi; dx++) {
                    if (buf[dx + x][y] == true) {
                        allfree = false;
                        break;
                    }
                }
                if (!allfree) {
                    continue;
                }
                // there's enough space => try it.
                for (int dx = 0; dx < wi; dx++) {
                    for (int dy = 0; dy < hi; dy++) {
                        buf[dx + x][dy + y] = true;
                    }
                }
                a[i]--;
                if (tryFill(x + wi, y)) {
                    return true;
                }
                for (int dx = 0; dx < wi; dx++) {
                    for (int dy = 0; dy < hi; dy++) {
                        buf[dx + x][dy + y] = false;
                    }
                }
                a[i]++;

            }
        }
        return false;
    }

    public static void main(String[] args) {
        for (int i = 0; i < 101; i++) {
            buf[i] = new boolean[101];
            for (int j = 0; j < 101; j++) {
                buf[i][j] = false;
            }
        }
        Scanner scan = new Scanner(System.in);
        W = scan.nextInt();
        H = scan.nextInt();
        c = scan.nextInt();
        long totalArea = 0;
        for (int i = 0; i < c; i++) {
            a[i] = scan.nextInt();
            w[i] = scan.nextInt();
            h[i] = scan.nextInt();
            totalArea += a[i]*w[i] * h[i];
        }
        if (totalArea >= W*H && tryFill(0,0)) {
            System.out.println("yes");
        } else {
            System.out.println("no");
        }
    }
}

File: 2018summer/07.12/solutions_2015/carpets/submissions/accepted/thorsten.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cassert>

using namespace std;

#define FOR(i,a,b) for(int i = (a); i < (b); i++)
#define DBGDO(X) cerr << #X << " = " << (X) << endl

struct Carpet {
    int count;
    int w;
    int h;
};

int rw, rh;
int cnt;
Carpet cps[120];

bool room[500][500];

bool tryfill(int x, int y) {
    /*
     * imagine the room:
     *        x
     *   ------>
     *  |
     *  |
     * y|
     *  V
     *
     *  the (x,y) is the topmost (and among those the leftmost) free cell
     */
    // wrap lines around
    if (x >= rw) return tryfill(0,y+1);
    // if we left the room already there is nothing left to fill
    if (y >= rh) return true;
    // if cell was lift already go to next one
    if (room[x][y] == true) return tryfill(x+1, y);
    // for any carpet
    FOR (c, 0, cnt) if (cps[c].count > 0) {
        Carpet* cur = cps + c;
        int curw = cur->w;
        int curh = cur->h;
        // in any rotation
        FOR (asdf,0,2) {
            swap(curw,curh);
            // check that it does not leave the room boundaries
            if (x + curw > rw) continue;
            if (y + curh > rh) continue;
            bool allfree = true;
            // check that the occupied cells aren't occupied already by other carpets
            FOR (cx, x, curw + x) {
                if (!allfree) break;
                FOR (cy, y, y+1) {
                    if (room[cx][cy] == true) {
                        allfree = false;
                        break;
                    }
                }
            }
            if (!allfree) continue;
            // if all are still free, place carpet
            cur->count--;
            FOR (cx, x, curw + x) FOR (cy, y, curh + y) {
                    assert(room[cx][cy] == false);
                    room[cx][cy] = true;
            }
            // continue filling
            if (tryfill(x+curw,y))
                return true;
            // if it failed, move carpet back to stock
            cur->count++;
            FOR (cx, x, curw + x) FOR (cy, y, curh + y) {
                    assert(room[cx][cy] == true);
                    room[cx][cy] = false;
            }
            // don't try the other rotation for squares
            if (curw == curh) break;
        }
    }
    return false;
}

int main() {
    cin >> rw >> rh;
    cin >> cnt;
    int A = 0;
    FOR (i,0,cnt) {
        cin >> cps[i].count >> cps[i].w >> cps[i].h;
        A += cps[i].count * cps[i].w * cps[i].h;
    }
    FOR (i, 0, rw) FOR (j, 0, rh) {
        room[i][j] = false;
    }
    if (A >= rw*rh && tryfill(0,0)) cout << "yes" << endl;
    else cout << "no" << endl;
    return 0;
}
